### ISSUE TITLE
Provide a setting to automatically exclude all gems

### ISSUE THREAD

**Issue by gmcabrita** [Shopify/ruby-lsp#2660]
### I have checked that this feature is not already implemented

- [x] This feature does not exist

### Use case

The performance of the global go to symbol in ruby-lsp is noticeably worse when not excluding any gems than when excluding every gem.

Here are some short videos showing the difference:
| All gems ignored | No gems ignored|
|------------------|----------------|
| https://github.com/user-attachments/assets/c5d5d922-df3b-46c7-b9f7-6dd65c9359cc | https://github.com/user-attachments/assets/ece1805e-0cba-4592-a51c-cf57f9c66a6e |

Currently, if I want to exclude every gem, I'm forced to list them all in my editor's config file: https://github.com/gmcabrita/dotfiles/blob/f7b8d85449b4f6da3acbc7428548ceb4ca29f999/.config/zed/settings.json#L137-L585.

I also have to update that list as time goes on and new gems are added to an existing project, or I jump to a new project that has gems I am not currently ignoring.

Providing a setting to exclude all gems without having to list them would be a nice improvement.

### Description

Possibly support one of these in the editor settings, ensuring that no gems are ever processed/indexed:
```json
{
  "excludedGemPatterns": ["*"]
}
```

or 

```
{
  "excludeAllGems": true
}
```

**Comment by gmcabrita** [comment]
> If I understand it correctly, `global go to symbol` is equivalent to workspace symbols in VS Code. 

Yes, this is for the `workspace/symbol` method.

> If that's the case, we already started excluding dependencies from the request with/without indexing them (https://github.com/Shopify/ruby-lsp/pull/2131). Can you please upgrade to the latest `ruby-lsp` to confirm it?

I checked with v0.20.0 and the performance difference is still pretty much the same.

> Also, excluding all gems from indexing will seriously degrade the functionality of other features like go-to-definition, autocompletion...etc. so it's not something we'll encourage users to do and it's rarely the right solution to the issue.

I understand that this degrades the go-to definition and autocomplete functionality for external gems. Personally, the performance hit of including all these gems is so annoying that I don't care if I have autocomplete on them or not. I suspect other developers are affected as well, they probably just don't care about editor latency as much as I do or they would have excluded all the gems like I did.

I understand if this is not a setting that `ruby-lsp` wants to support. Performance for `workspace/symbol` should still be worked on even if this setting is supported, since it's just a band-aid.

**Comment by st0012** [comment]
I agree that we can possibly improve the performance of the `workspace/symbol` request so it has less delay. Therefore, I'll change the title of this issue to be explicit about this end goal.

But I also need to be honest that given our team's current backlog, this will be low on our list for at least a while. That said, if you want to investigate what's causing the slowdown, PRs are welcome 🙂

**Comment by janko** [comment]
I too would like a way to exclude indexing of all gems. That not only speeds up indexing significantly, but also `workspace/symbol` searches as the OP said, which I rarely use in my project primarily due to slowness (the other reason is https://github.com/Shopify/ruby-lsp/issues/3781). I never wish to `cmd-click` to a gem symbol or search a gem's methods, I always open gem codebases separately if I want to learn how they work, I don't think I have any use of gem codebases being indexed.

**Comment by janko** [comment]
I measured that a `workspace/symbol` search is **3.5x faster** in my Rails project when gems are **not indexed**. Coupled with Zed's lack of debouncing while typing the search query and Ruby LSP's single-threadedness, the performance difference really adds up.

**Comment by pkondzior** [comment]
I think there are low hanging fruits that can be addressed here without removing ruby gems symbols from the index. The `workspace/symbol` is using ruby implemented Jaro-Winkler algorithm, switching to something C based would  definitively improve the performance. on huge projects. Swtiching to either `amatch` could speed up 9x and to `jaro_winkler` 29x

```ruby
require 'benchmark'
require 'did_you_mean'   # DidYouMean::JaroWinkler.distance (distance: 0..1, lower = closer)
require 'amatch'         # Amatch::JaroWinkler (similarity: 0..1, higher = closer)
require 'jaro_winkler'   # ::JaroWinkler (similarity)

# Alias to avoid namespace conflicts
JWGem = ::JaroWinkler

STRING_PAIRS = [
  ['kitten',    'sitting'],
  ['distance',  'instance'],
  ['jaro',      'jarowinkler'],
  ['rails',     'ruby'],
  ['developer', 'developper'],
  ['microsoft', 'microsofft'],
  ['benchmark', 'benchmarck'],
  ['abcdefg',   'abcxefg'],
  ['abcdefg',   'uvwxyz'],
  ['openai',    'openAIL']
]

N = 100_000

Benchmark.bm(32) do |x|
  x.report('DidYouMean::JaroWinkler (dist)') do
    N.times do
      STRING_PAIRS.each do |a, b|
        DidYouMean::JaroWinkler.distance(a, b)
      end
    end
  end

  x.report('Amatch::JaroWinkler (sim)') do
    N.times do
      STRING_PAIRS.each do |a, b|
        Amatch::JaroWinkler.new(a).match(b)
      end
    end
  end

  x.report('jaro_winkler gem (sim)') do
    N.times do
      STRING_PAIRS.each do |a, b|
        JWGem.similarity(a, b)
      end
    end
  end
end
``` 

```
DidYouMean::JaroWinkler (dist)     3.532733   0.024145   3.556878 (  3.557902)
Amatch::JaroWinkler (sim)          0.388356   0.007825   0.396181 (  0.396189)
jaro_winkler gem (sim)             0.137851   0.001409   0.139260 (  0.139265)
```

**Comment by janko** [comment]
@pkondzior Yeah, the performance could be improved by adding a new dependency. However, Zed seems to filter out any symbols that aren't in the project from the search results anyway (that includes external gems), so I could also just reduce the work Ruby LSP needs to do in my case.

**Comment by pkondzior** [comment]
@janko I think they are already ignored from the results https://github.com/Shopify/ruby-lsp/blob/main/lib/ruby_lsp/requests/workspace_symbol.rb#L29 but I am working on profiling it right now and I see there could be some caching involved, there is not that much gain from JaroWinkler external dependecy thoo

**Comment by janko** [comment]
Oh, didn't realize it was ignored at the Ruby LSP level. Wondering if it makes sense to first filter out symbols not in workspace, and *then* perform fuzzy search on the symbols that remain. Sounds like time could be saved by avoiding fuzzy searching results that will be skipped anyway 🤔

**Comment by pkondzior** [comment]
I was able to cut down by 57% time for the first call of `workspace/Symbol` method and 70% for consecutive calls on `ruby-lsp` project directory, I think on bigger projects this numbers would get even higher.  All this without even applying your proposed idea @janko I think with that we could get even further with cutting time profile on that call.

I will be committing changes into https://github.com/Shopify/ruby-lsp/pull/3784 stay tuned.

As a sneak preview the most consuming right now are:
1. Constant resolution requires comments to be parsed upfront during entry initialisation, this could happen lazily and be cached
2. Entry is not caching information about its dependency role, this could be cached
3. Uri file_paths are being treated for escaping/unscaping with windows semantics on non windows platforms (ie, ruby-lsp always look for C: or windows path separators on non windows platforms), we could have windows and non windows versions of escaping/unscaping for Uris which would be much more robuust on given platform than they are right now. 
4. Constant resolution happens always when we are performing fuzzy search, this is extra cruft  I've added in my PR, I've solved it  by moving constant resolution at end of the indexing phase, this could be backported to other response builders semantics and speed them up too.

Addditonally I think 2 and 3 could be cached when entries are being created during indexing, this would make first call to `workspace/Symbol` more robust because right now even with caching on the method level we are still putting this pressure on the response builder, but both depdenecy role and file_path for the uri could be both cached during entry initailization.

**Comment by janko** [comment]
@pkondzior I also worked on optimizing this, using the approach I mentioned – filtering first, fuzzy matching after – see https://github.com/Shopify/ruby-lsp/pull/3792. I was able to get a **4x speedup** in my medium-sized Rails application at work, going from ~0.8s to ~0.2s 🤘🏻 

It could be faster if we stored the fact whether an URI is coming from a dependency inside the URI, since we know it at build time, which I believe is what you meant by in 2️⃣. However, I consider this a good start, as I believe this additional optimization would add coupling to `URI::Generic`.

I'd be curious to know more about the constant resolution overhead, as it didn't pop up when I profiled using Stackprof.

**Comment by pkondzior** [comment]
I am not sure if I follow your code changes, seems like you added some additional indirection without actually applying changes you ware talking about, the cut probably comes from using raw_path and omitting parser, but is still going to be calculated every time you are performing symbol search which is not neccessary imho, the path is not going to change so why check it on entry every time, also what is the condition on fuzzy search doing ? and why introduce that indirection ?

**Comment by pkondzior** [comment]
@janko I do not support the changes you proposed with the raw_path, this seems to be messy and create coupling that affects only fuzzy_search, here are the changes I've been talking about [earlier](https://github.com/Shopify/ruby-lsp/issues/2660#issuecomment-3446270636) that improve performance by 2.7 times  vs main branch

https://github.com/Shopify/ruby-lsp/pull/3784

**Comment by janko** [comment]
> seems like you added some additional indirection without actually applying changes you ware talking about

That's incorrect. I flipped the order of file filtering and fuzzy matching, just like I had proposed.

> do not support the changes you proposed with the raw_path

Then let's discuss that on the PR. This issue should be used for discussing high-level approaches to speeding things up, not technical details of one proposed solution.

> here are the changes I've been talking about https://github.com/Shopify/ruby-lsp/issues/2660#issuecomment-3446270636 that improve performance by 2.7 times vs main branch

I recommend not conflating bug fixes with performance improvements, to make PRs easier to review and safer to merge. I'm eagerly awaiting your bugfix to be merged 🤞🏻

**Comment by pkondzior** [comment]
> I recommend not conflating bug fixes with performance improvements, to make PRs easier to review and safer to merge. I'm eagerly awaiting your bugfix to be merged 🤞🏻

My bugfix requires constant and method resolution to be in place before fuzzy search happens so I believe it is required. If not done right it will regress on perfrmance.

I can extract URI and Entry performance improvements into separate PR if that is necessary, I think they have good impact not only on workspace symbol generation but also in other space like definition resolution. 

> It could be faster if we stored the fact whether an URI is coming from a dependency inside the URI, since we know it at build time, which I believe is what you meant by in 2️⃣. 

This is my point, I think the dependency check can be generic and on the entry level, this can affect many other places, your change really introduces something that will omit the main code path and not provide much more benefits in other places.

Also if the original change regarding fuzzy search doesn't bring improvement without it why introduce the indirection ? The problem lies in the URI and how it handles different platforms.

**Comment by janko** [comment]
> Also if the original change regarding fuzzy search doesn't bring improvement without it why introduce the indirection ? The problem lies in the URI and how it handles different platforms.

Because together with skipping calls to `URI::Generic#to_standardized_path` during workspace symbol search it brings significant speedup. I could easily fix that bottleneck (by skipping computation), and without it the fuzzy search is the bottleneck, which I couldn't easily fix (at least without introducing a new dependency with potential C extensions).

You keep calling it "indirection", but it's different logic. Let's benchmark indexing and symbol search speeds in our respective PRs and see where we stand.

**Comment by vinistock** [comment]
Just to bring context from the other issue since the discussion got spread around multiple: our team is in the middle of working in a redesign of the indexer process for better performance and correctness, which should help improve workspace symbols.

For now, please don't put major efforts into changing our current architecture, as everything under `lib/ruby_indexer` will be completely substituted by the new work.
